<a href="https://colab.research.google.com/github/leinaxd/Tps/blob/master/a4/a4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GitHub

In [ ]:
#@markdown SSH

#@markdown Ingrese directorio con contraseña privada
directorio = "" #@param {type:"string"}

!ssh-keygen -h dsd
# !ssh-keygen -t ed25519 -C "leinaxd@gmail.com"
# %cd /root/.ssh/
# !ls -la
# !cat id_ed25519.pub

In [1]:
#@markdown Cargar
%cd ~
%cd /content/

!git clone https://github.com/leinaxd/Tps/

%cd Tps/a4

/root
/content
Cloning into 'Tps'...
remote: Enumerating objects: 700, done.
remote: Counting objects: 100% (700/700), done.
remote: Compressing objects: 100% (524/524), done.
remote: Total 700 (delta 391), reused 412 (delta 163), pack-reused 0
Receiving objects: 100% (700/700), 78.37 MiB | 23.67 MiB/s, done.
Resolving deltas: 100% (391/391), done.
Checking out files: 100% (221/221), done.
/content/Tps/a4


In [7]:
AUTH = "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" #@param {type:"string"}
!git config --global user.email "leinaxd@gmail.com"
!git config --global user.name "leinaxd"
!git remote set-url origin https://$AUTH@github.com/leinaxd/Tps.git

In [8]:
commit = "a4 1g: nmt trained" #@param {type:"string"}

!git add .
!git commit -m "$commit"
!git push



[master 41fcec3] a4 1g: nmt trained
 5 files changed, 2 insertions(+), 2 deletions(-)
Counting objects: 9, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (9/9), done.
Writing objects: 100% (9/9), 12.59 KiB | 12.59 MiB/s, done.
Total 9 (delta 4), reused 0 (delta 0)
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/leinaxd/Tps.git
   9918213..41fcec3  master -> master


In [ ]:
#@markdown pull / update
!git pull

# Instalar librerías

In [3]:
!pip install sentencepiece
!pip install sacrebleu

     |████████████████████████████████| 1.2 MB 5.2 MB/s 
     |████████████████████████████████| 90 kB 4.4 MB/s 


# CS224N Assignment 4: Neural Machine Translation with RNNs (45 + 30 Points)

Esta tarea se divide en dos partes: *Neural Machine Translation with RNNs* y *Analyzing NMT systems*. El primero es básicamente codificación y el segundo son preguntas escritas. Si te atoras en la primera sección, puedes trabajar en la segunda ya que las secciones son independientes.
La notación e implementación de los *sistemas NMT* es tramposa. Recomendamos la lectura de Zhang, 2020. https:arxic.org/abs/2010.04791 para entender mejor la tarea de traducción *Cherokee-Inglés* que sirvió de inspiración para este proyecto.

## 1. Neural Machine Translation with RNNs (45 points)

En *Machine Translation* el objetivo es convertir una oración desde un lenguaje *source*(Cherokee) hacia un lenguaje *target* (Inglés).<br>
En este proyecto implementaremos un modelo neuronal **seq2seq** con **Atención**. En esta sección, describiremos el **procedimiento de entrenamiento** para el sistema **NMT** propuesto, el cual utiliza un **Encoder** **LSTM** **Bidireccional** y un **Decoder** **LSTM** **Unidireccional**

### Descripción del modelo (Procedimiento de aprendizaje)
![modelo](https://github.com/leinaxd/Tps/raw/master/a4/imgs/arquitecturaNMT.jpg)
Figura 1. Modelo seq2seq con Atención multiplicativa, mostrada en la tercera fase del decoder. Los estados ocultos $h_i^{enc}$ y los estados de las celdas $c_i^{enc}$ se definen a continuación.

Dada una oración en un lenguaje de origen, buscaremos los embeddings a nivel *subword* a partir de la matriz de *embeddings* obteniendo $x_1...x_m \hspace{10mm} (x_i\in \mathbb{R}^{e\times 1})$, donde $m$ es la longitud de la oración de entrada y $e$ la dimensión del embedding.
Alimentamos estos embeddings al encoder bidireccional, obteniendo estados ocultos y estados de celda tanto para el *forwards* como *backwards* del **LSTM**. Las versiones *Forwards* y *Backwards* se concatenan para obtener el estado oculto $\mathbf{h_i^{enc}}$ y estado de celda $\mathbf{c_i^{enc}}$:

$\begin{align}
\mathbf{h_i^{enc}} &= [\overleftarrow{\mathbf{h_i^{enc}}}; \overrightarrow{\mathbf{h_i^{enc}}}] \hspace{10mm} &&con \hspace{5mm} \mathbf{h_i^{enc}}\in \mathbb{R}^{2h\times 1},\hspace{5mm} \overleftarrow{\mathbf{h_i^{enc}}},\overrightarrow{\mathbf{h_i^{enc}}}\in \mathbb{R}^{h\times 1} \hspace{5mm} 1\le i\le m\\
\mathbf{c_i^{enc}} &= [\overleftarrow{\mathbf{c_i^{enc}}}; \overrightarrow{\mathbf{c_i^{enc}}}] \hspace{10mm} &&con \hspace{5mm} \mathbf{c_i^{enc}}\in \mathbb{R}^{2h\times 1},\hspace{5mm} \overleftarrow{\mathbf{c_i^{enc}}},\overrightarrow{\mathbf{c_i^{enc}}}\in \mathbb{R}^{h\times 1} \hspace{5mm} 1\le i\le m\\
\end{align} $

Luego inicializaremos el primer estado oculto del **Decoder** $\mathbf{h_0^{dec}}$ y el estado de celda $\mathbf{c_0^{dec}}$ con una proyección lineal del estado oculto y celda final del **encoder**.<br>
(Si no es obvio, piense por qué consideramos $[\overleftarrow{\mathbf{h_1^{enc}}},\overrightarrow{\mathbf{h_m^{enc}}}]$ el estado final del encoder.)

$\begin{align}
\mathbf{h_0^{dec}} &= \mathbf{W_h} [\overleftarrow{\mathbf{h_1^{enc}}}; \overrightarrow{\mathbf{h_m^{enc}}}] \hspace{10mm} &&con \hspace{5mm} \mathbf{h_0^{dec}}\in \mathbb{R}^{h\times 1},\hspace{5mm} \mathbf{W_h}\in \mathbb{R}^{h\times 2h}\\
\mathbf{c_0^{dec}} &= \mathbf{W_c} [\overleftarrow{\mathbf{c_1^{enc}}}; \overrightarrow{\mathbf{c_m^{enc}}}] \hspace{10mm} &&con \hspace{5mm} \mathbf{c_0^{dec}}\in \mathbb{R}^{h\times 1},\hspace{5mm} \mathbf{W_c}\in \mathbb{R}^{h\times 2h} \\
\end{align} $

Con el **Decoder** inicializado, debemos alimentarlo con la oración *target* (destino). En el $t^{mo}$ paso, buscamos el embedding para la $t^{mo}$ *subword* $\mathbf{y_t}\in \mathbb{R}^{e\times 1}$. Luego concatenamos $\mathbf{y_t}$ con un vector de salida combinado $o_{t-1}\in \mathbb{R}^{h\times 1}$ del instante anterior (Explicado más adelante), para producir $\bar{\mathbf{y_t}}\in \mathbb{R}^{(e+h)\times 1}$. Notar que para el primer *subword* destino (*target*) (El token de inicio) $\mathbf{o_0}$ es un vector de ceros. Luego alimentamos $\bar{\mathbf{y_t}}$ como entrada al decoder.

$\mathbf{h_t^{dec}, \mathbf{c_t^{dec}}} = Decoder(\bar{\mathbf{y_t}}, \mathbf{h_{t-1}^{dec}}, \mathbf{c_{t-1}^{dec}}) \hspace{10mm} donde\hspace{5mm} \mathbf{h_t^{dec}}\in \mathbb{R}^{h\times 1}, \mathbf{c_t^{dec}}\in \mathbb{R}^{h\times 1}
$

Luego utlizaremos $\mathbf{h_t^{dec}}$ para calcular la atención de forma multiplicativa sobre $\mathbf{h_1^{enc} \dots h_m^{enc}}$:

$\begin{align} \mathbf{e_{t,i}}
&=(\mathbf{h_t^{dec}})^T \mathbf{W_{attProj}}\mathbf{h_i^{enc}}
\hspace{10mm} &&donde
\hspace{5mm} \mathbf{e_t}\in \mathbb{R}^{m\times 1},
\hspace{5mm} \mathbf{W_{attProj}}\in \mathbb{R}^{h\times 2h}
\hspace{5mm} 1\le i \le m\\
\alpha_t&=softmax(\mathbf{e_t})
\hspace{10mm} &&donde
\hspace{5mm} \alpha_t\in \mathbb{R}^{m\times 1}\\
\mathbf{a_t}&=\sum_{i=1}^{m}\alpha_{t,i}\  \mathbf{h_i^{enc}}
\hspace{10mm} &&donde
\hspace{5mm} \mathbf{a_t}\in \mathbb{R}^{2h\times 1}
\end{align}$

Ahora concatenaremos la salida de atención $\mathbf{a_t}$ con el estado oculto del decoder $\mathbf{h_t^{dec}}$ y pasarlo por una capa lineal, $tanh$ y una capa *dropout* para atender la salida combinada $\mathbf{o_t}$

$\begin{align}
\mathbf{u_t} &= [\mathbf{a_t};\mathbf{h_t^{dec}}]
\hspace{10mm} &&donde
\hspace{5mm} \mathbf{u_t}\in \mathbb{R}^{3h\times 1}
\hspace{5mm}\\
\mathbf{v_t} &=
\mathbf{W_u}\ \mathbf{u_t}
\hspace{10mm} &&donde
\hspace{5mm}\mathbf{v_t}\in \mathbb{R}^{h\times 1},
\hspace{5mm}\mathbf{W_u}\in \mathbb{R}^{h\times 3h}\\
\mathbf{o_t} &=dropout(tanh(\mathbf{v_t}))
\hspace{10mm} &&donde
\hspace{5mm} \mathbf{o_t}\in \mathbb{R}^{h\times 1}
\end{align}$

Luego generamos la distribución de proababilidad $\mathbf{P_t}$ sobre los *subwords* destino (*target*) al
 $t^{mo}$ paso.

$\mathbf{P_t} = softmax(\mathbf{W_{vocab}\ o_t})
\hspace{10mm} donde\hspace{5mm}
\mathbf{P_t}\in \mathbb{R}^{\mathbf{V_t}\times 1},
\hspace{5mm}\mathbf{W_{vocab}}\in \mathbb{R}^{\mathbf{V_t}\times h}$

Aquí $\mathbf{V_t}$ es el tamaño del vocabulario destino (*target*). Finalmente, para entrenar la red podemos calcular la entropía cruzada softmax entre $\mathbf{P_t}$ y $\mathbf{g_t}$, donde $\mathbf{g_t}$ es el vector *one-hot* del *subword* destino al instante $t$

$J_t (\theta ) =CrossEntropy(\mathbf{P_t},\mathbf{g_t})$

Donde $\theta$ representa todos los parámetros del modelo y $J_t(\theta ) es el costo en el instante $t$ del **decoder**. Ahora que hemos descripto el modelo, tratemos de implementarlo para la traducción Cherokee-Inglés!

### Configurando la máquina virtual
Siga las instrucciones de https://docs.google.com/document/d/1BQOAjhBxWbywkB4rMFH9iinb6YHSjaWw1TOVIGfyYho para crear una copia de la máquina virtual. Esto debería tomarte unos 45 minutos. Sin embargo necesitarás la GPU para entrenar el modelo, recomendamos fuertemente que primero desarrolles el código localmente y asegurate que funciona antes de intentar entrenarlo en la máquina virtual. El tiempo de la GPU es limitado y costoso. Toma aproximadamente entre 30 minutos y 1 hora entrenar el sistema NMT. Finalmente, asegurate de apagar la máquina virtual cuando no la estés utilizando.

Si la subscripción a Azure se queda sin dinero, tu máquina virtual será temporalmente bloqueada e inaccesible. Si esto sucede, por favor complete el siguiente formulario https://docs.google.com/forms/d/e/1FAIpQLSdpd5CgwSulB_wGCKQb_VmLscicTswiTVBxnw40xVzeuS4BKQ/viewform

Para ejecutar el código en tu máquina local, ejecuta los siguientes comandos para crear el apropiado entorno
`conda env create --file local_env.yml`
Notar que este entorno virtual no se necesita en la máquina virtual.

### Implementación y preguntas escritas

#### a. Implemente la función `pad_sents` en `utils.py` que producirá estas oraciones tabuladas (2 points)
Para aplicar las operaciones sobre los tensores, tenemos que asegurarnos que todas las oraciones del mismo *batch* tengan la misma longitud.
* Identifique la oración más larga del *batch* y tabule a las demás en dicha longitud.

#### b. (3 points) Implemente la función `__init__` de `model_embeddings.py` 
Que inicializa los *embeddings* del *source* (oración de origen) y *target* (oración destino)<br>

#### c. (4 points) Implemente la función `__init__` de `nmt_model.py` para inicializar los embeddings del modelo y las capas (LSTM, projection y dropout) del sistema NMT
(Ademas Utiliza la clase `ModelEmbeddings` de `model_embeddings.py`)
#### d. (8 points) Implementar la función `encode` de `nmt_model.py`.
Esta función 
* Convierte las oraciones *source* ya tabuladas en un tensor $\mathbf{X}$
* Genera $\mathbf{h_1^{enc}\dots h_m^{enc}}$
* Calcula el estado inicial $\mathbf{h_0^{dec}}$ y celda inicial $\mathbf{c_0^{dec}}$ del **decoder**.

Puedes ejecutar una verificación no exhaustiva mediante:
`python sanity_check.py 1d`

In [4]:
!python sanity_check.py 1d

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
Running Sanity Check for Question 1d: Encode
--------------------------------------------------------------------------------
enc_hiddens Sanity Checks Passed!
dec_init_state[0] Sanity Checks Passed!
dec_init_state[1] Sanity Checks Passed!
--------------------------------------------------------------------------------
All Sanity Checks Passed for Question 1d: Encode!
--------------------------------------------------------------------------------


#### e. (8 points) Implemente la función `decode` de `nmt_model.py`.
Esta función construye $\bar{\mathbf{y_t}}$ y ejecuta la función `step` por cada índice de tiempo de la entrada.
Puedes ejecutar un test no exhaustivo mediante: `python sanity_check.py 1e`

In [5]:
!python sanity_check.py 1e

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
--------------------------------------------------------------------------------
Running Sanity Check for Question 1e: Decode
--------------------------------------------------------------------------------
torch.Size([23, 5, 3])
combined_outputs Sanity Checks Passed!
--------------------------------------------------------------------------------
All Sanity Checks Passed for Question 1e: Decode!
--------------------------------------------------------------------------------


#### f. (10 points) Implemente la función `step` de `nmt_model.py`.
Esta función aplica la celda LSTM del **Decoder** para un único índice de tiempo, calcula el encoding del *subword* destino $\mathbf{h_t^{dec}}$, los puntajes de atención $\mathbf{e_t}$, la distribución de atención $\alpha_t$, la salidas de atención $\mathbf{a_t}$ y finalmente la salida combinada $\mathbf{o_t}$.
Puedes ejecutar un test no exhaustivo mediante: `python sanity_check.py 1f`

In [6]:
!python sanity_check.py 1f

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
--------------------------------------------------------------------------------
Running Sanity Check for Question 1f: Step
--------------------------------------------------------------------------------
dec_state[0] Sanity Checks Passed!
dec_state[1] Sanity Checks Passed!
combined_output  Sanity Checks Passed!
e_t Sanity Checks Passed!
--------------------------------------------------------------------------------
All Sanity Checks Passed for Question 1f: Step!
--------------------------------------------------------------------------------


#### g. (3 points) La función `generate_sent_masks()` de `nmt_model.py` produce un tensor llamado `enc_masks`. Con dimensiones (batch_size, max source sentence length) y contiene unos en las posiciones correspondientes a los símbolos *pad* de entrada y ceros para el resto de símbolos. Observar cómo se utilizan las máscaras durante el cálculo de la atención en la función `step`.
Primero explique (aprox. 3 oraciones) qué efecto tienen las máscaras sobre el mecanismo completo de atención. Luego explique (en una o dos oraciones), por qué es necesario utilizar las máscaras de esta manera.

<font color = "green">

* Al procesar en **Batch** se toma un conjunto de **b** 
oraciones, todas ellas tabuladas a la **misma longitud**.

* Se utiliza una **máscara** para que no tome en cuenta los instantes de **tiempo** que la RNN calcula el estado $\mathbf{h_i}$ para la entrada **< pad>**
 
* Está implementado poniendo a cero el coeficiente que pondera a determinado estado. ($\alpha_{t,i} = 0$)

$$a_t = \sum_{i=1}^m \alpha_{t,i} h_i^{enc}$$
</font>

### Ejecución

Es momento de ejecutar las cosas! Ejecute lo siguiente para generar el archivo de vocabulario necesario
`sh run.sh vocab`
Ó si estás en un entorno Windows, utilice el siguiente comando. Asegurese que ejecutas esto en un entorno que tiene a python como PATH.
`run.bat vocab`

Como se observó anteriormente, recomendamos que desarrolles el código en tu ordenador personal. Confirma que estás ejecutando el entorno conda correcto y luego ejecuta el siguiente comando para entrenar el modelo en tu máquina local.

(linux) `sh run.sh train_local`
(windows) `run.bat train_local`

In [ ]:
# !sh run.sh vocab
!sh run.sh train_local

Deberías ver una disminución significativa del coste durante las iteraciones iniciales. Una vez que garantices que tu código no se rompe (ej. ejecutarlo hasta la iteración 10 o 20) enciende la máquina virtual desde el portal de Azure.
 Luego lee la sección *Managing Code Deployment to a VM` de nuestra guía práctica a VMs https://docs.google.com/document/d/1jtANWXbIYXMZO_2X7jupauPxcEbz-TVJkdatg4gzOdk/edit
Por instrucciones de cómo subir tu código a la VM.

Luego instala los paquetes necesarios a tu máquina virtual
`pip install -r gpu_requirements.txt`

In [9]:
!pip install -r gpu_requirements.txt

     |████████████████████████████████| 46 kB 2.0 MB/s 
  Attempting uninstall: tqdm
    Found existing installation: tqdm 4.62.0
    Uninstalling tqdm-4.62.0:
      Successfully uninstalled tqdm-4.62.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spacy 2.2.4 requires tqdm<5.0.0,>=4.38.0, but you have tqdm 4.29.1 which is incompatible.
panel 0.12.1 requires tqdm>=4.48.0, but you have tqdm 4.29.1 which is incompatible.
fbprophet 0.7.1 requires tqdm>=4.36.1, but you have tqdm 4.29.1 which is incompatible.


Finalmente, ir a la sección de la guía y sigue las instrucciones para crear una nueva sesión `tmux`. Concretamente, ejecutar el siguiente comando para crear una sesión `tmux` llamada `nmt`

`tmux new -s nmt`

Una vez configurado tu VM y estás en una sesión `tmux`, ejecute:
(linux) sh run.sh train
(Windows) run.bat train

Una vez sepas que tu código funciona correctamente, puedes desconectarte de sesión y cerrar tu conexión ssh al servidor.
Para hacerlo ejecuta:
`tmux detach`

Puedes volver a tu modelo de entrenamiento reconectandote al servidor y añadiendo la sesión `tmux` con:
`tmux a -t nmt`

In [10]:
!sh run.sh train

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
uniformly initialize parameters [-0.100000, +0.100000]
use device: cuda:0
begin Maximum Likelihood training
epoch 1, iter 10, avg. loss 209.71, avg. ppl 3507.98 cum. examples 320, speed 1395.33 words/sec, time elapsed 5.89 sec
epoch 1, iter 20, avg. loss 178.38, avg. ppl 780.96 cum. examples 640, speed 1372.37 words/sec, time elapsed 12.14 sec
epoch 1, iter 30, avg. loss 165.85, avg. ppl 562.57 cum. examples 960, speed 1322.42 words/sec, time elapsed 18.47 sec
epoch 1, iter 40, avg. loss 171.09, avg. ppl 484.10 cum. examples 1280, speed 1303.92 words/sec, time elapsed 25.27 sec
epoch 1, iter 50, avg. loss 162.76, avg. ppl 379.93 cum. examples 1600, speed 1439.92 words/sec, time elapsed 31.36 sec
epoch 1, iter 60, avg. loss 154.27, avg. ppl 354.72 cum. examples 1920, speed 1428.34 words/sec, time elapsed 37.24 sec
epoch 1, iter 70, avg. loss 154.14, avg. ppl 309.02 cum. example

#### h. (3 points) Una vez tu modelo haya finalizado de entrenar (debería tomar menos de 1 hora en la máquina virtual) ejecute el siguiente comando para testear el modelo.
(linux) sh run.sh test
(windows) run.bat test
Por favor, reporte el puntaje BLEU. Debería ser mayor a 10.


In [11]:
!sh run.sh test

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
load test source sentences from [./chr_en_data/test.chr]
load test target sentences from [./chr_en_data/test.en]
load model from model.bin
Decoding:   0% 0/1000 [00:00<?, ?it/s]/usr/local/lib/python3.7/dist-packages/torch/_tensor.py:575: UserWarning: floor_divide is deprecated, and will be removed in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values.
To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor'). (Triggered internally at  /pytorch/aten/src/ATen/native/BinaryOps.cpp:467.)
  return torch.floor_divide(self, other)
Decoding: 100% 1000/1000 [01:56<00:00,  8.61it/s]
Corpus BLEU: 12.125998270142949


<font color = "green">
BLEU = 12.125998
</font>

#### i. (4 points) (written)
En clase hemos aprendido sobre la atención de producto interno, atención multiplicativa y atención aditiva. Como recordatorio, la atención de producto interno es $\mathbf{e_{t,i}}=\mathbf{s_t^T\ h_i}$, La atención multiplicativa es $\mathbf{e_{t,i}}=\mathbf{s_t^T\ W\ h_i}$ y la atención aditiva es $\mathbf{e_{t,i}}=\mathbf{v^T}\ tanh(\mathbf{W_1\ h_i}+\mathbf{W_2\ s_t})$

* i. (2 points) Explique una ventaja y una desventaja del producto interno como atención comparado con la atención multiplicativa.
* ii. (2 points) Explicar una ventaja y una desventaja de la atención aditiva respecto a la atención multiplicativa

## 2. Analizando el sistema NMT (30 points)

En la parte 1 hemos modelado nuestro problema NMT a un *subword-level*. Esto es, dada una oración en el idioma de origen, encontramos los componentes a nivel *subword* desde la matriz de *embeddings*.
Alternativamente podríamos haber modelado el problema NMT a nivel palabra, obervando palabras completas desde la matriz de *embeddings*.
### a. Por qué modelar el sistema **NMT** bajo *subword-level* vs *word-level*? (2 points)
**Pista**: El Cherokee es un idioma polysintético

<font color ="green">

</font>

### b. Por qué las codificaciones (*Embeddings*) a nivel *letra* y *Subword* suelen ser más chicos que la codificación a nivel *palabra*? (2 points)
Dar una razón en 1-2 oraciones.

<font color ="green">

</font>

### c. Cómo ayuda el entrenamiento multilenguaje en mejorar el rendimiento sobre lenguajes de bajos recursos? (2 points)
Un reto de entrenar modelos **NMT** exitosos es la carencia de datos del lenguaje, particularmente para lenguajes de recursos escasos como el Cherokee. Una forma de responder a este reto es mediante el entrenamiento multi-lenguaje, donde entrenamos nuestro modelo **NMT** sobre varios idiomas (incluyendo Cherokee). <br>
Puedes leer más en https://ai.googleblog.com/2019/10/exploring-massively-multilingual.html

<font color ="green">

</font>

### d. (6 points)
Aquí presentamos 3 ejemplos de errores encontrados en las salidas de nuestro modelo **NMT** (que es el mismo entrenado por tí). Los errores están subrayados.
Por cada ejemplo de la oración de origen, referenciar la traducción a inglés y la traducción obtenida por el modelo.
* Incluir las posibles razones que expliquen el error cometido por el modelo. (Tanto como su construcción lingüística o una limitación específica del modelo)
* Describir una posible manera para corregir el error observado. Hay más de una manera de corregir un error. Por ejemplo, se podría ajustar el tamaño de la capa oculta o cambiando el mecanismo de atención.

A continuación están las traducciones a analizar. Sólo analizar el error subrayado de cada oración. Se asegura que no tienes que saber Cherokee para responder estas preguntas. Pero si quieres saber un poco más de este lenguaje, sientete libre de consultar https://www.cherokeedictionary.net/ para buscar palabras.

#### i. (2 points)
**Origen de la traducción** (Source) = *Yona utsesdo anitsilvsgi digvtanv uwoduisdei* <br>
**Traducción de referencia** *Fern had a crown of daisies in her hair*<br>
**Traducción del modelo**: *Fern had <u>her hair</u> with her hair*
#### ii. (2 points)
**Origen de la traducción** (Source) = *Ulihelisdi nigalisda.* <br>
**Traducción de referencia** *She is very excited.*<br>
**Traducción del modelo**: *<u>it's</u> joy*
#### iii. (2 points)
**Origen de la traducción** (Source) = *Tsesdi hana yitsadawoesdi usdi atsadi!* <br>
**Traducción de referencia** *Don't swim there, Littlefish!*<br>
**Traducción del modelo**: *Don't know how <u>a small fish!</u>*

### e. (4 points)
Exploremos las salidas del modelo que hemos entrenado! Las traducciones del set de testeo producidas en el punto 1-i deberían encontrarse en
 `outputs/test_outputs.txt`.
1. (2 points) Encontrar el límite donde las traducciones predichas son correctas para secuencias largas (4 o 5 palabras). Verificar el archivo destino *target* (Inglés).
Contiene el archivo de entrenamiento ese texto (casi) literal? Si lo tiene o no, qué aprendió el sistema a hacer?
2. (2 points) Encontrar el límite donde la traducción predicha comienza correcta durante una secuencia de 4 o 5 palabras, pero luego diverge (donde el final de la oración parece no tener sentido alguno). Qué dice esto sobre el comportamiento del decodificador?

### f. (14 points)
El puntaje **BLEU** es la métrica de evaluación automática más utilizada para sistemas **NMT**. Se suele calcular sobre el set de testeo completo, pero aquí consideraremos el puntaje para un solo ejemplo (*).
Suponga que tenemos una oración de origen $\mathbf{s}$, un conjunto de $k$ traducciones de referencia $\mathbf{r_1 \dots r_k}$ y una traducción candidata $\mathbf{c}$.
Para calcular el puntaje **BLEU** de $\mathbf{c}$, primero debemos calcular la precisión n-grama modificada $p_n$ de $\mathbf{c}$, para cada $n=1,2,3,4$ donde n es el n-grama.

$p_n=\frac{\sum_{ngram\in c} min(
max_{i=1...k}(Count_{\mathbf{r_i}} (n-gram), Count_{\mathbf{c}} (n-gram))
)}
{\sum_{ngram\in c} Count_{\mathbf{c}}(ngram)}$

Aquí por cada n-grama que aparece en la traducción candidata $\mathbf{c}$, contamos la máxima cantidad de veces que aparece en cualquiera traducción de referencia, limitado a la cantidad de veces que aparece en $\mathbf{c}$ (este es el numerador). Dividimos este número por la cantidad de n-gramas en $\mathbf{c}$ (denominador).

(\*) Esta definición de puntaje **BLEU** nivel oración coincide con la función `sentence_bleu()` en el paquete `nltk`. Observar que la función del NLTK es sensible a las mayúsculas. En esta pregunta, todo el texto está en minúsculas. https://www.nltk.org/api/nltk.translate.html#nltk.translate.bleu_score.sentence_bleu

Luego, calculamos la *Brevity pealty* **BP**. <br>
Sea $len(\mathbf{c})$ la longitud de $\mathbf{c}$ y sea $len(\mathbf{r})$ la longitud de la traducción de referencia que es más cercana a $len(\mathbf{c})$ (En el caso de dos referencias equidistantes, elegir $len(\mathbf{r})$ como la más corta.)

$ BP = \begin{cases}
1 \hspace{10 mm} &si \hspace{5mm} len(\mathbf{c})\ge len(\mathbf{r}) \\
exp(1-\frac{len(\mathbf{r})}{len (\mathbf{c})})  &e.o.c.
\end{cases} $

Finalmente, el puntaje **BLEU** para el candidato $\mathbf{c}$ respecto a $\mathbf{r_1 \dots r_n}$ es:

$BLEU = BP·exp(\sum_{n=1}^4 \lambda_n\ log\ p_{n})$

Donde $\lambda_1, \lambda_2, \lambda_3, \lambda_4$ son pesos cuya suma es 1. El $log$ es el logaritmo natural.

#### i. Considere el ejemplo en español. (5 points)
Oración de **Origen**: **El amor todo lo puede**<br>
Traducción de referencia $\mathbf{r_1}$:     *Love can always find a way* <br>
Traducción de referencia $\mathbf{r_2}$:     *Love makes anything possible*<br>
Modelo NMT $\mathbf{c_1}$: *The love can always do* <br>
Modelo NMT $\mathbf{c_2}$: *love can make anything possible*

* Calcule el puntaje **BLEU** para $\mathbf{c_1}$ y $\mathbf{c_2}$.
Sea $\lambda_i = 0.5$ para $i\in \{1,2\}$ y $\lambda_i = 0$ para $i\in \{3,4\}$ (Esto significa ignorar los 3-gramas y 4-gramas. no calcular $p_3$ y $p_4$)
Cuando se calculan los puntajes BLEU, mostrar tu trabajo (valores $p_1, p_2, len(\mathbf{c}),len(\mathbf{r}), BP$). Notar que los puntajes **BLEU** se pueden expresar entre 0 y 1, o entre 0 y 100. El código utiliza la escala 0-100, mientras que las preguntas utilizan la escala 0-1.

Cuál de las dos traducciones se considera mejor según **BLEU**? Estás de acuerdo en que es la mejor traducción?

ii. (5 points) Nuestro disco duro se ha corrompido y hemos perdido la traducción de referencia $\mathbf{r_2}$. Por favor, recalcule el puntaje **BLEU** para $\mathbf{c_1}$ y $\mathbf{c_2}$, esta vez solamente respecto a $\mathbf{r_1}$. Cuál de las dos traducciones recive un mayor puntaje? Estás de acuerdo en que es la mejor traducción?
iii. (2 points) Por la disponibilidad de datos, los sistemas NMT suelen evaluarse respecto a una sola traducción de referencia. Explique por qué podría ser problemático.
iv. (2 points) Enumerar dos ventajas y dos desventajas de **BLEU** comparado con la evaluación humana como métrica para *Machine Translation*